src/compose/images.ts

```diff
@@ -68,8 +67,92 @@ const imageFetchLastFailureTime: Dictionary<ReturnType<
 	typeof process.hrtime
 >> = {};
 const imageCleanupFailures: Dictionary<number> = {};
+
+type ImageState = Pick<Image, 'status' | 'downloadProgress'>;
+type ImageTask = {
+	// Indicates whether the task has been finished
+	done?: boolean;
+
+	// Current image state of the task
+	context: Image;
+
+	// Update the task with new context. This is a pure function
+	// meaning it doesn't modify the original task
+	update: (change?: ImageState) => ImageTaskUpdate;
+
+	// Finish the task. This is a pure function
+	// meaning it doesn't modify the original task
+	finish: () => ImageTaskUpdate;
+};
+
+type ImageTaskUpdate = [ImageTask, boolean];
+
+// Create new running task with the given initial context
+function createTask(initialContext: Image) {
+	// Task has only two state, is either running or finished
+	const running = (context: Image): ImageTask => {
+		return {
+			context,
+			update: ({ status, downloadProgress }: ImageState) =>
+				// Keep current state
+				[
+					running({
+						...context,
+						...(status && { status }),
+						...(downloadProgress && { downloadProgress }),
+					}),
+					// Only mark the task as changed if there is new data
+					[status, downloadProgress].some((v) => !!v),
+				],
+			finish: () => [finished(context), true],
+		};
+	};
+
+	// Once the task is finished, it cannot go back to a running state
+	const finished = (context: Image): ImageTask => {
+		return {
+			done: true,
+			context,
+			update: () => [finished(context), false],
+			finish: () => [finished(context), false],
+		};
+	};
+
+	return running(initialContext);
+}
+
+const runningTasks: { [imageId: number]: ImageTask } = {};
+function reportEvent(event: 'start' | 'update' | 'finish', state: Image) {
+	const { imageId } = state;
+
+	let emitChange = true;
+
+	// Get the current task and update it in memory
+	const currentTask =
+		event === 'start' ? createTask(state) : runningTasks[imageId];
+	runningTasks[imageId] = currentTask;
+
+	// TODO: should we assert that the current task exists at this point?
+	// On update, update the corresponding task with the new state if it exists
+	if (event === 'update' && currentTask) {
+		const [updatedTask, changed] = currentTask.update(state);
+		runningTasks[imageId] = updatedTask;
+		emitChange = changed;
+	}
+
+	// On update, update the corresponding task with the new state if it exists
+	if (event === 'finish' && currentTask) {
+		[, emitChange] = currentTask.finish();
+		delete runningTasks[imageId];
+	}
+
+	if (emitChange) {
```

Comment:
if (emitChange && currentTask)

Should we allow change events if there happens to be an undefined currentTask?

Reply:
Good point, probably not